### ISSUE TITLE
feat(semver): support ranges for workspace dependencies

### ISSUE THREAD

**Issue by ecraig12345** [JamieMason/syncpack#124]
## Description

Unless I'm missing something, syncpack only seems to support exact versions for workspace dependencies.

Sometimes it would be helpful to have the option of allowing a range, but verifying that the range matches the current version. Some example scenarios:
- Specify dev deps on the repo's internal tooling package(s) as `*`
- Specify deps on repo packages as `^x.y.z` (this can reduce duplication for consumers if they depend on multiple packages from the repo)

The behavior with this setting enabled would be that given a workspace package with version `1.2.3`:
- Correct: consistently use `^1.0.0`, `~1.2.0`, `*`, etc
- Incorrect: `^0.9.0`, `^1.3.0`, `~1.1.0`, etc (or inconsistent semver spec)

## Suggested Solution

I'm not sure exactly where this should fit into the config. The most obvious thing I could think of (which I tried before filing this) is putting it in `semverGroups` like this:

```json
    "semverGroups": [
      {
        "range": "^",
        "dependencyTypes": ["workspace"],
        "packages": ["**"],
        "dependencies": ["**"]
      }
    ]
```

## Help Needed

I might be able to implement this if you point to the right place.

**Comment by JamieMason** [comment]
On an initial glance it maybe be enough just to remove this line https://github.com/JamieMason/syncpack/blob/2f1ffba70ccd27599780b2e602283755490087d6/src/get-context/get-groups/semver-group.ts#L19. It was an assumption I had that the version would/should always exactly match given the origin of that package is there in the same monorepo.

**Comment by ecraig12345** [comment]
Did you mean this line? https://github.com/JamieMason/syncpack/blob/2f1ffba70ccd27599780b2e602283755490087d6/src/get-context/get-groups/semver-group.ts#L19

I made that change locally and ran it against [this repo/branch](https://github.com/ecraig12345/lage/tree/syncpack) (with some further updates to the config), and still got errors.

This is the config I used. It also gave the exact same results without the `semverGroups`.
```
  "syncpack": {
    "dependencyTypes": [
      "dev",
      "prod",
      "peer",
      "workspace"
    ],
    "semverGroups": [
      {
        "range": "^",
        "dependencyTypes": ["workspace"],
        "packages": ["**"],
        "dependencies": ["**"]
      }
    ]
  }
```

Sample of the errors:
```
✘ @lage-run/cache 0.4.3 is developed in this repo at packages/cache/package.json
  ^0.4.3 in dependencies of packages/cli/package.json
  ^0.4.3 in dependencies of packages/scheduler/package.json
✘ @lage-run/cli 0.10.1 is developed in this repo at packages/cli/package.json
  ^0.10.1 in dependencies of packages/e2e-tests/package.json
  ^0.10.1 in devDependencies of packages/lage/package.json
```

**Comment by JamieMason** [comment]
No problem I'll take a look

**Comment by JamieMason** [comment]
I've done some work on this, I'm on holiday next week but the next release will include it. 

I've been thinking that `workspace:` should be treated as a new kind of semver range but I need to check that over – I noticed recently in fluid framework (I test against that repo locally) that you have `workspace:~` which I hadn't seen before, so the `~` portion of that would need some special handing adding.

```diff
- export type SemverRange = '' | '*' | '>' | '>=' | '.x' | '<' | '<=' | '^' | '~';
+ export type SemverRange = '' | '*' | '>' | '>=' | '.x' | '<' | '<=' | '^' | '~' | 'workspace:';
```

**Comment by ecraig12345** [comment]
Thanks! I guess I should have clarified that by "workspace dependencies" I assumed it meant just "packages defined within the same workspace." I haven't personally used the `workspace:` syntax.

**Comment by jonathanj** [comment]
> I've been thinking that `workspace:` should be treated as a new kind of semver range but I need to check that over – I noticed recently in fluid framework (I test against that repo locally) that you have `workspace:~` which I hadn't seen before, so the `~` portion of that would need some special handing adding.
> 
> ```diff
> - export type SemverRange = '' | '*' | '>' | '>=' | '.x' | '<' | '<=' | '^' | '~';
> + export type SemverRange = '' | '*' | '>' | '>=' | '.x' | '<' | '<=' | '^' | '~' | 'workspace:';
> ```

I generally use `workspace:*` within my workspace packages, although the [Yarn docs on the topic](https://yarnpkg.com/features/workspaces#workspace-ranges-workspace) say that the other ranges work too (but recommend the use of `workspace:*`).

My syncpack config ended up ignoring `myorg/**` packages and dependencies because I couldn't get syncpack to recognise and treat them as workspace local dependencies.

**Comment by JamieMason** [comment]
> Thanks! I guess I should have clarified that by "workspace dependencies" I assumed it meant just "packages defined within the same workspace." I haven't personally used the `workspace:` syntax.
> – @ecraig12345

My bad, I keep confusing the various things "workspace" can refer to. Ok, I think I have a fix for this locally but it's not everything you're looking for. It will allow ranges to be used when you depend on a package which is developed locally, but those range versions will need to be identical to each other as per current syncpack behaviour. At a later date though hopefully https://github.com/JamieMason/syncpack/issues/128 can deliver some support for ranges being satisfied.

> I generally use `workspace:*` within my workspace packages, although the [Yarn docs on the topic](https://yarnpkg.com/features/workspaces#workspace-ranges-workspace) say that the other ranges work too (but recommend the use of `workspace:*`).
> – @jonathanj

I'm tempted to limit support to only `workspace:*` now you mention it, at least in the first drop.

> My syncpack config ended up ignoring `myorg/**` packages and dependencies because I couldn't get syncpack to recognise and treat them as workspace local dependencies.
> – @jonathanj

Can you create an issue for this scenario so it can be looked into?

Thanks both

**Comment by JamieMason** [comment]
Released in [syncpack@10.0.0](https://github.com/JamieMason/syncpack/releases/tag/10.0.0), please let me know if you run into any issues.
